### DISCUSSION TITLE
Why is loss scaled by dp group size before calling backward ?

### DISCUSSION THREAD

**Discussion by sappenn** [NVIDIA-NeMo/Automodel#1403] [Q&A] [answered]
Hi, I am trying to understand why is loss scaled by the DP group size https://github.com/NVIDIA-NeMo/Automodel/blob/v0.3.0rc4/nemo_automodel/recipes/llm/train_ft.py#L1223 instead of simply `loss.backward()`

My setup is FSDP2 + TP=8 on 16 nodes. So in total there are 16 data parallel ranks with the FSDP2 mesh. And within a node, TP=8. Model is Qwen3-8B dense model.
In the preceding section, where loss is calculated, it it calculated from the logits which is sharded over the TP mesh. And there is a `logits.full_tensor` operation performed where logits are gathered from all TP ranks within a node. This is within the loss function  https://github.com/NVIDIA-NeMo/Automodel/blob/v0.3.0rc4/nemo_automodel/components/loss/masked_ce.py#L77. This makes sense that all TP ranks within a node have the same logits value. 

From this value, local_loss is calculated. As soon as loss's backward is called on the loss, gradients are synced between dp ranks on the FSDP2 mesh.  This operation is similar to FSDP2 operation where only the backward of local loss is taken into consideration. Then why loss is again scaled by the dp_group size ?

Any pointers appreciated.

**Comment by hemildesai** [comment, marked as answer]
Hi, this is because FSDP averages the gradients by default during the reduction phase ([ref](https://github.com/pytorch/pytorch/blob/v2.10.0/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L741-L742)) so we need to scale the loss by `dp_group_size` since our loss is calculated [per token](https://github.com/NVIDIA-NeMo/Automodel/blob/r0.3.0/nemo_automodel/recipes/llm/train_ft.py#L1225) across the global batch.
